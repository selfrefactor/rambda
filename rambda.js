export * from './src/F'
export * from './src/T'
export * from './src/add'
export * from './src/adjust'
export * from './src/all'
export * from './src/allPass'
export * from './src/always'
export * from './src/and'
export * from './src/any'
export * from './src/anyPass'
export * from './src/append'
export * from './src/applySpec'
export * from './src/assoc'
export * from './src/assocPath'
export * from './src/both'
export * from './src/clone'
export * from './src/complement'
export * from './src/compose'
export * from './src/concat'
export * from './src/cond'
export * from './src/curry'
export * from './src/clamp'
export * from './src/dec'
export * from './src/defaultTo'
export * from './src/difference'
export * from './src/dissoc'
export * from './src/divide'
export * from './src/drop'
export * from './src/dropLast'
export * from './src/either'
export * from './src/endsWith'
export * from './src/equals'
export * from './src/filter'
export * from './src/find'
export * from './src/findIndex'
export * from './src/flatten'
export * from './src/flip'
export * from './src/forEach'
export * from './src/fromPairs'
export * from './src/groupBy'
export * from './src/groupWith'
export * from './src/has'
export * from './src/head'
export * from './src/identical'
export * from './src/identity'
export * from './src/ifElse'
export * from './src/inc'
export * from './src/includes'
export * from './src/indexBy'
export * from './src/indexOf'
export * from './src/init'
export * from './src/intersperse'
export * from './src/intersection'
export * from './src/is'
export * from './src/isEmpty'
export * from './src/isNil'
export * from './src/join'
export * from './src/keys'
export * from './src/last'
export * from './src/lastIndexOf'
export * from './src/length'
export * from './src/lens'
export * from './src/lensIndex'
export * from './src/lensPath'
export * from './src/lensProp'
export * from './src/map'
export * from './src/match'
export * from './src/mathMod'
export * from './src/max'
export * from './src/maxBy'
export * from './src/mean'
export * from './src/median'
export * from './src/merge'
export * from './src/min'
export * from './src/minBy'
export * from './src/modulo'
export * from './src/multiply'
export * from './src/negate'
export * from './src/none'
export * from './src/not'
export * from './src/nth'
export * from './src/omit'
export * from './src/over'
export * from './src/partial'
export * from './src/path'
export * from './src/paths'
export * from './src/pathOr'
export * from './src/pick'
export * from './src/pickAll'
export * from './src/pipe'
export * from './src/pluck'
export * from './src/prepend'
export * from './src/product'
export * from './src/prop'
export * from './src/propEq'
export * from './src/propIs'
export * from './src/propOr'
export * from './src/range'
export * from './src/reduce'
export * from './src/reject'
export * from './src/repeat'
export * from './src/replace'
export * from './src/reverse'
export * from './src/set'
export * from './src/slice'
export * from './src/sort'
export * from './src/sortBy'
export * from './src/split'
export * from './src/splitEvery'
export * from './src/startsWith'
export * from './src/subtract'
export * from './src/symmetricDifference'
export * from './src/sum'
export * from './src/tail'
export * from './src/take'
export * from './src/takeLast'
export * from './src/tap'
export * from './src/testMethod'
export * from './src/times'
export * from './src/toLower'
export * from './src/toPairs'
export * from './src/toString'
export * from './src/toUpper'
export * from './src/transpose'
export * from './src/trim'
export * from './src/type'
export * from './src/uniq'
export * from './src/uniqWith'
export * from './src/update'
export * from './src/values'
export * from './src/view'
export * from './src/without'
export * from './src/zip'
export * from './src/xor'
export * from './src/zipObj'
