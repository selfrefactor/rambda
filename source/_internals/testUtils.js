import combinate from 'combinate'
import {
  equals,
  filter,
  forEach,
  init,
  last,
  omit,
  switcher,
  type,
} from 'rambdax'

const omitOk = omit('ok')

const PENDING = 'PENDING'
const RESULTS_EQUAL = 'results are equal'
const RESULTS_MISMATCH = 'results are different'
const ERRORS_EQUAL = 'errors are equal'
const ERRORS_MESSAGE_MISMATCH = 'errors messages are different'
const ERRORS_TYPE_MISMATCH = 'errors types are different'
const SHOULD_THROW = 'Rambda should throw'
const SHOULD_NOT_THROW = 'Rambda should not throw'
const ALL_ERROR_LABELS = {
  RESULTS_MISMATCH,
  ERRORS_MESSAGE_MISMATCH,
  ERRORS_TYPE_MISMATCH,
  SHOULD_NOT_THROW,
  SHOULD_THROW,
}
const MISSING = 'Missing error handle in parseError'

function parseError(err){
  const typeError = switcher(err)
    .is(x => x instanceof TypeError, 'TypeError')
    .is(x => x instanceof SyntaxError, 'SyntaxError')
    .is(x => x instanceof RangeError, 'RangeError')
    .is(x => x instanceof Error, 'Error')
    .default(MISSING)

  if (typeError === MISSING){
    throw new Error('typeError === MISSING')
  }

  return {
    message : err.message,
    type    : typeError,
    ok      : true,
  }
}

function executeSync(
  fn, inputs, returnsFunctionFlag
){
  let result = PENDING
  let error = { ok : false }
  try {
    result = returnsFunctionFlag ?
      fn(...init(inputs))(last(inputs)) :
      fn(...inputs)
  } catch (e){
    error = parseError(e)
  }

  return {
    result,
    error,
  }
}

async function executeAsync(
  fn, inputs, returnsFunctionFlag
){
  let result = PENDING
  let error = { ok : false }
  try {
    result = returnsFunctionFlag ?
      await fn(...init(inputs))(last(inputs)) :
      await fn(...inputs)
  } catch (e){
    error = parseError(e)
  }

  return {
    result,
    error,
  }
}

export function profileMethod({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  returnsFunctionFlag = false,
  fn,
}){
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)

    test(getTestTitle(...inputs), () => {
      const { result, error } = executeSync(
        fn, inputs, returnsFunctionFlag
      )

      expect({
        result,
        error : error.ok ? omitOk(error) : PENDING,
        inputs,
      }).toMatchSnapshot()
    })
  })
}

export function profileMethodAsync({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  returnsFunctionFlag = false,
  fn,
}){
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)
    test(getTestTitle(...inputs), async () => {
      const { result, error } = await executeAsync(
        fn,
        inputs,
        returnsFunctionFlag
      )

      expect({
        result,
        error : error.ok ? omitOk(error) : PENDING,
        inputs,
      }).toMatchSnapshot()
    })
  })
}

export function compareToRamda(
  fn, fnRamda, returnsFunctionFlag
){
  return (...inputs) => {
    const { result, error } = executeSync(
      fn, inputs, returnsFunctionFlag
    )
    const { result: ramdaResult, error: ramdaError } = executeSync(
      fnRamda,
      inputs,
      returnsFunctionFlag
    )

    const toReturn = {
      result,
      ramdaResult,
      ramdaError : ramdaError.ok ? omitOk(ramdaError) : PENDING,
      error      : error.ok ? omitOk(error) : PENDING,
    }

    if (result !== PENDING){
      if (ramdaError.ok){
        return {
          ...toReturn,
          ok    : false,
          label : SHOULD_THROW,
        }
      }

      if (equals(result, ramdaResult)){
        return {
          ...toReturn,
          ok    : true,
          label : RESULTS_EQUAL,
        }
      }

      return {
        ...toReturn,
        ok    : false,
        label : RESULTS_MISMATCH,
      }
    }

    if (equals(error, ramdaError)){
      return {
        ...toReturn,
        ok    : true,
        label : ERRORS_EQUAL,
      }
    }

    if (ramdaError.ok){
      return {
        ...toReturn,
        ok    : false,
        label :
          ramdaError.type === error.type ?
            ERRORS_MESSAGE_MISMATCH :
            ERRORS_TYPE_MISMATCH,
      }
    }

    return {
      ...toReturn,
      ok    : false,
      label : SHOULD_NOT_THROW,
    }
  }
}

export const getTestTitle = (...inputs) => inputs.map(type).join(' | ')

export const compareCombinations = ({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  returnsFunctionFlag = false,
  setCounter = () => {},
  callback = x => {},
  fn,
  fnRamda,
}) => {
  const counter = {
    RESULTS_MISMATCH        : 0,
    SHOULD_THROW            : 0,
    SHOULD_NOT_THROW        : 0,
    ERRORS_TYPE_MISMATCH    : 0,
    ERRORS_MESSAGE_MISMATCH : 0,
    TOTAL_TESTS             : 0,
  }

  const increaseCounter = comparedResult => {
    if (comparedResult.ok) return
    let counterProp
    forEach((x, prop) => {
      if (x === comparedResult.label) counterProp = prop
    }, ALL_ERROR_LABELS)
    counter[ counterProp ]++
  }

  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)
  const compareOutputs = compareToRamda(
    fn, fnRamda, returnsFunctionFlag
  )

  afterAll(() => callback(counter))

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)

    test(getTestTitle(...inputs), () => {
      const compared = compareOutputs(...inputs)
      setCounter()
      if (!compared.ok){
        // if (compared.label === RESULTS_MISMATCH){
        increaseCounter(compared)
        expect({
          ...compared,
          inputs,
        }).toMatchSnapshot()
      }
      counter.TOTAL_TESTS++
    })
  })
}

const demoMap = new Map()

demoMap.set('a', 1)
demoMap.set('b', 2)

const demoWeakMap = new WeakMap()

demoWeakMap.set({ a : 0 }, 1)
demoWeakMap.set({ b : 0 }, 2)

const demoWeakSet = new WeakSet()
const foo = { a : 1 }
const bar = { b : 2 }

demoWeakSet.add(foo)
demoWeakSet.add(bar)

const demoBigInt = BigInt(9007199254740991)

const demoRegExp = new RegExp('ab+c', 'i')
const demoSymbol = Symbol('foo')

export const FALSY_VALUES = [ null, undefined, false, NaN, '' ]
export const EXTRA_BUILD_IN_OBJECTS = [
  Infinity,
  -Infinity,
  Symbol,
  new ArrayBuffer(8),
  demoRegExp,
  demoMap,
  demoWeakMap,
  demoWeakSet,
  demoBigInt,
  demoSymbol,
]
