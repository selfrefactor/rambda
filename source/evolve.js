import { mapObject } from './mapObject.js'
import { type } from './type.js'

export function evolve(rules) {
  return mapObject((x, prop) => type(rules[prop]) === 'Function' ? rules[prop](x): x)
}
