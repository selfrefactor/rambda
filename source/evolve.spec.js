import { evolve as evolveRamda } from 'ramda'

import { add } from '../rambda.js'
import { compareCombinations, compareToRamda } from './_internals/testUtils.js'
import { evolve } from './evolve.js'

test('happy', () => {
  const rules = {
    foo    : add(1),
    nested : { bar : x => Object.keys(x).length },
  }
  const input = {
    a      : 1,
    foo    : 2,
    nested : { bar : { z : 3 } },
  }
  const result = evolve(rules, input)
  expect(result).toEqual({
    a      : 1,
    foo    : 3,
    nested : { bar : 1 },
  })
})

test('nested rule is wrong', () => {
  const rules = {
    foo    : add(1),
    nested : { bar : 10 },
  }
  const input = {
    a      : 1,
    foo    : 2,
    nested : { bar : { z : 3 } },
  }
  const result = evolve(rules)(input)
  expect(result).toEqual({
    a      : 1,
    foo    : 3,
    nested : { bar : { z : 3 } },
  })
})

test('is recursive', () => {
  const rules = {
    nested : {
      second : add(-1),
      third  : add(1),
    },
  }
  const object = {
    first  : 1,
    nested : {
      second : 2,
      third  : 3,
    },
  }
  const expected = {
    first  : 1,
    nested : {
      second : 1,
      third  : 4,
    },
  }
  const result = evolve(rules, object)
  expect(result).toEqual(expected)
})

test('ignores primitive values', () => {
  const rules = {
    n : 2,
    m : 'foo',
  }
  const object = {
    n : 0,
    m : 1,
  }
  const expected = {
    n : 0,
    m : 1,
  }
  const result = evolve(rules, object)
  expect(result).toEqual(expected)
})

test('with array', () => {
  const rules = [ add(1), add(-1) ]
  const list = [ 100, 1400 ]
  const expected = [ 101, 1399 ]
  const result = evolve(rules, list)
  expect(result).toEqual(expected)
})

const rulesObject = { a : add(1) }
const rulesList = [ add(1) ]
const possibleIterables = [ null, undefined, '', 42, [], [ 1 ], { a : 1 } ]
const possibleRules = [ ...possibleIterables, rulesList, rulesObject ]

describe('brute force', () => {
  compareCombinations({
    firstInput : possibleRules,
    callback   : errorsCounters => {
      expect(errorsCounters).toMatchInlineSnapshot(`
        {
          "ERRORS_MESSAGE_MISMATCH": 0,
          "ERRORS_TYPE_MISMATCH": 4,
          "RESULTS_MISMATCH": 0,
          "SHOULD_NOT_THROW": 51,
          "SHOULD_THROW": 0,
          "TOTAL_TESTS": 63,
        }
      `)
    },
    secondInput : possibleIterables,
    fn          : evolve,
    fnRamda     : evolveRamda,
  })
})
