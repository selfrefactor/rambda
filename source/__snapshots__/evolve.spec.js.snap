// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`brute force Array | Null 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [],
    null,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": null,
  "result": "PENDING",
}
`;

exports[`brute force Array | Null 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [
      1,
    ],
    null,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": null,
  "result": "PENDING",
}
`;

exports[`brute force Array | Null 3`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [
      [Function],
    ],
    null,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": null,
  "result": "PENDING",
}
`;

exports[`brute force Array | Number 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [],
    42,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": 42,
  "result": "PENDING",
}
`;

exports[`brute force Array | Number 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [
      1,
    ],
    42,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": 42,
  "result": "PENDING",
}
`;

exports[`brute force Array | Number 3`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [
      [Function],
    ],
    42,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": 42,
  "result": "PENDING",
}
`;

exports[`brute force Array | Object 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [],
    {
      "a": 1,
    },
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": {
    "a": 1,
  },
  "result": "PENDING",
}
`;

exports[`brute force Array | Object 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [
      1,
    ],
    {
      "a": 1,
    },
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": {
    "a": 1,
  },
  "result": "PENDING",
}
`;

exports[`brute force Array | Object 3`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [
      [Function],
    ],
    {
      "a": 1,
    },
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": {
    "a": 1,
  },
  "result": "PENDING",
}
`;

exports[`brute force Array | String 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [],
    "",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": "",
  "result": "PENDING",
}
`;

exports[`brute force Array | String 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [
      1,
    ],
    "",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": "",
  "result": "PENDING",
}
`;

exports[`brute force Array | String 3`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [
      [Function],
    ],
    "",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": "",
  "result": "PENDING",
}
`;

exports[`brute force Array | Undefined 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [],
    undefined,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": undefined,
  "result": "PENDING",
}
`;

exports[`brute force Array | Undefined 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [
      1,
    ],
    undefined,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": undefined,
  "result": "PENDING",
}
`;

exports[`brute force Array | Undefined 3`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    [
      [Function],
    ],
    undefined,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": undefined,
  "result": "PENDING",
}
`;

exports[`brute force Null | Array 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    null,
    [],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Null | Array 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    null,
    [
      1,
    ],
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading '0')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | Null 1`] = `
{
  "error": {
    "message": "'iterable' and 'rules' are from wrong type Null",
    "type": "Error",
  },
  "inputs": [
    null,
    null,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": null,
  "result": "PENDING",
}
`;

exports[`brute force Null | Number 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    null,
    42,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": 42,
  "result": "PENDING",
}
`;

exports[`brute force Null | Object 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    null,
    {
      "a": 1,
    },
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of null (reading 'a')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Null | String 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    null,
    "",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": "",
  "result": "PENDING",
}
`;

exports[`brute force Null | Undefined 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    null,
    undefined,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": undefined,
  "result": "PENDING",
}
`;

exports[`brute force Number | Array 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    42,
    [],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Number | Array 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    42,
    [
      1,
    ],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [
    1,
  ],
  "result": "PENDING",
}
`;

exports[`brute force Number | Null 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    42,
    null,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": null,
  "result": "PENDING",
}
`;

exports[`brute force Number | Number 1`] = `
{
  "error": {
    "message": "'iterable' and 'rules' are from wrong type Number",
    "type": "Error",
  },
  "inputs": [
    42,
    42,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": 42,
  "result": "PENDING",
}
`;

exports[`brute force Number | Object 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    42,
    {
      "a": 1,
    },
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": {
    "a": 1,
  },
  "result": "PENDING",
}
`;

exports[`brute force Number | String 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    42,
    "",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": "",
  "result": "PENDING",
}
`;

exports[`brute force Number | Undefined 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    42,
    undefined,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": undefined,
  "result": "PENDING",
}
`;

exports[`brute force Object | Array 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    {
      "a": 1,
    },
    [],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Object | Array 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    {
      "a": [Function],
    },
    [],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Object | Array 3`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    {
      "a": 1,
    },
    [
      1,
    ],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [
    1,
  ],
  "result": "PENDING",
}
`;

exports[`brute force Object | Array 4`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    {
      "a": [Function],
    },
    [
      1,
    ],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [
    1,
  ],
  "result": "PENDING",
}
`;

exports[`brute force Object | Null 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    {
      "a": 1,
    },
    null,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": null,
  "result": "PENDING",
}
`;

exports[`brute force Object | Null 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    {
      "a": [Function],
    },
    null,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": null,
  "result": "PENDING",
}
`;

exports[`brute force Object | Number 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    {
      "a": 1,
    },
    42,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": 42,
  "result": "PENDING",
}
`;

exports[`brute force Object | Number 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    {
      "a": [Function],
    },
    42,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": 42,
  "result": "PENDING",
}
`;

exports[`brute force Object | String 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    {
      "a": 1,
    },
    "",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": "",
  "result": "PENDING",
}
`;

exports[`brute force Object | String 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    {
      "a": [Function],
    },
    "",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": "",
  "result": "PENDING",
}
`;

exports[`brute force Object | Undefined 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    {
      "a": 1,
    },
    undefined,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": undefined,
  "result": "PENDING",
}
`;

exports[`brute force Object | Undefined 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    {
      "a": [Function],
    },
    undefined,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": undefined,
  "result": "PENDING",
}
`;

exports[`brute force String | Array 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    "",
    [],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force String | Array 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    "",
    [
      1,
    ],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [
    1,
  ],
  "result": "PENDING",
}
`;

exports[`brute force String | Null 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    "",
    null,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": null,
  "result": "PENDING",
}
`;

exports[`brute force String | Number 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    "",
    42,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": 42,
  "result": "PENDING",
}
`;

exports[`brute force String | Object 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    "",
    {
      "a": 1,
    },
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": {
    "a": 1,
  },
  "result": "PENDING",
}
`;

exports[`brute force String | String 1`] = `
{
  "error": {
    "message": "'iterable' and 'rules' are from wrong type String",
    "type": "Error",
  },
  "inputs": [
    "",
    "",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": "",
  "result": "PENDING",
}
`;

exports[`brute force String | Undefined 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    "",
    undefined,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": undefined,
  "result": "PENDING",
}
`;

exports[`brute force Undefined | Array 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    undefined,
    [],
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": [],
  "result": "PENDING",
}
`;

exports[`brute force Undefined | Array 2`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    undefined,
    [
      1,
    ],
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of undefined (reading '0')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Undefined | Null 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    undefined,
    null,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": null,
  "result": "PENDING",
}
`;

exports[`brute force Undefined | Number 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    undefined,
    42,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": 42,
  "result": "PENDING",
}
`;

exports[`brute force Undefined | Object 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    undefined,
    {
      "a": 1,
    },
  ],
  "label": "errors types are different",
  "ok": false,
  "ramdaError": {
    "message": "Cannot read properties of undefined (reading 'a')",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Undefined | String 1`] = `
{
  "error": {
    "message": "iterableType !== rulesType",
    "type": "Error",
  },
  "inputs": [
    undefined,
    "",
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": "",
  "result": "PENDING",
}
`;

exports[`brute force Undefined | Undefined 1`] = `
{
  "error": {
    "message": "'iterable' and 'rules' are from wrong type Undefined",
    "type": "Error",
  },
  "inputs": [
    undefined,
    undefined,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": undefined,
  "result": "PENDING",
}
`;
